8300208: Optimize Adler32 stub for AVX-512 targets.

jatin-bhateja opened the pull request:
Patch optimizes Adler32 stub for AVX512 target.

Main computation loop now uses zero extended lane widening load vector operation.

New sequence also honors AVX3Thresholds so that implementation uses existing AVX2 instruction sequence on relevant targets
if input size is smaller than threshold limit (default 4096).

Following are the result of an [existing JMH micro ](https://github.com/openjdk/jdk/blob/master/test/micro/org/openjdk/bench/java/util/TestAdler32.java)on various targets.

**System Configurations : Turbo frequency scaling is disabled, all the data is collected at fixed frequency of 2.8 GHz.
SUT1   : Intel® Xeon® Platinum 8480+ Processor (Sapphire Rapids)  56C 2S
SUT2   : Intel(R) Xeon(R) Platinum 8380 CPU (Icelake Server) 40C 2S
SUT3   : Intel(R) Xeon(R) Platinum 8280 CPU (Cascadelake Server) 28C 2S**


![image](https://user-images.githubusercontent.com/59989778/212934730-68717a61-191f-4dba-8c83-2eddf6007a47.png)

![image](https://user-images.githubusercontent.com/59989778/212934945-cada95ad-c93c-487f-bacc-928a2e3b5c21.png)

![image](https://user-images.githubusercontent.com/59989778/212935059-511aca3b-c736-40a2-bff6-89caf0664828.png)


Please review and share your feedback.

Best Regards,
Jatin

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300208](https://bugs.openjdk.org/browse/JDK-8300208): Optimize Adler32 stub for AVX-512 targets.


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12045/head:pull/12045` \
`$ git checkout pull/12045`

Update a local copy of the PR: \
`$ git checkout pull/12045` \
`$ git pull https://git.openjdk.org/jdk pull/12045/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12045`

View PR using the GUI difftool: \
`$ git pr show -t 12045`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12045.diff">https://git.openjdk.org/jdk/pull/12045.diff</a>

</details>

vnkozlov reviewed (commented):
Looks good to me. Let me test it.

vnkozlov reviewed (approved):
My testing passed. You need second review.

turbanoff commented on test/micro/org/openjdk/bench/java/util/TestAdler32.java:
```diff
@@ -27,20 +27,16 @@
     private byte[] bytes;
 
-    @Param({"64", "128", "256", "512", /* "1024", */ "2048", /* "4096", "8192", */ "16384", /* "32768", */ "65536"})
+    @Param({"64", "128", "256", "512", "1024",  "2048", "5012", "8192", "16384", "32768", "65536"})
```
nit:
```suggestion
    @Param({"64", "128", "256", "512", "1024", "2048", "5012", "8192", "16384", "32768", "65536"})
```

jatin-bhateja commented:
/label remove core-libs

jatin-bhateja commented:
Hi @sviswa7, can you kindly be second reviewer.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp:
```diff
@@ -132,28 +139,72 @@ address StubGenerator::generate_updateBytesAdler32() {
-  __ cmpptr(data, end);
-  __ jcc(Assembler::below, SLOOP1A);
+  if (VM_Version::supports_avx512vl()) {
+    __ cmpl(s, VM_Version::avx3_threshold());
+    __ jcc(Assembler::belowEqual, SPRELOOP1A_AVX2);
+    // AVX2 performs better for smaller inputs because of leaner post loop reduction sequence..
+    __ cmpl(s, 128);
+    __ jcc(Assembler::belowEqual, SPRELOOP1A_AVX2);
```
These two compares can be merged into one compare with larger of avx3_threshold() or 128.

sviswa7 commented on src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp:
```diff
@@ -132,28 +139,72 @@ address StubGenerator::generate_updateBytesAdler32() {
+    __ vpxor(yb, yb, yb, Assembler::AVX_512bit);
+
+    __ bind(SLOOP1A_AVX3);
+      __ evpmovzxbd(ydata, Address(data, 0), Assembler::AVX_512bit);
+      __ vpaddd(ya, ya, ydata, Assembler::AVX_512bit);
+      __ vpaddd(yb, yb, ya, Assembler::AVX_512bit);
+      __ addptr(data, CHUNKSIZE);
+      __ cmpptr(data, end);
```
This still processes 16 bytes worth of data in one loop iteration as the AVX2 loop. Have you given thoughts on processing double the size with AVX3?

jatin-bhateja replied:
If we double the size of inputs processed within the loop from 16 to 32 then subsequent ADLER_SCALE will also change from 0-31, given that we cannot accommodate more than 16 integer in a 512 bit vector hence it will not be possible to scale up the loop block size.

sviswa7 replied:
Could be done something like below:
[adler32.patch](https://github.com/openjdk/jdk/files/10521481/adler32.patch)

jatin-bhateja replied:
Yes, one level of unrolling will save loop jump back costs, I have incorporated this in the patch.

sviswa7 commented:
Could you please also update the test/hotspot/jtreg/compiler/intrinsics/zip/TestAdler32.java to throw Exception on failure?

jatin-bhateja commented:
Hi @sviswa7 , your comments have been addressed.

sviswa7 commented:
@jatin-bhateja There are some windows build failures. Please take a look.

sviswa7 reviewed (approved):
Looks good to me.

vnkozlov reviewed (changes requested):
This needs to be tested before integration since stub code was changes since v00.

jatin-bhateja commented:
> This needs to be tested before integration since stub code was changes since v00.

Hi @vnkozlov , please let me know once your test runs are through.

vnkozlov reviewed (approved):
My testing of latest version passed.

jatin-bhateja commented:
/integrate

vnkozlov commented:
@jatin-bhateja this changes seem cause regression https://bugs.openjdk.org/browse/JDK-8302358
Please, look.
